spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+  private final Map<Integer, PartitionSpec> specs;
+  private final Map<String, String> summary;
+  private final int defaultParallelism;
+  private final boolean snapshotIdInheritanceEnabled;
+  private final long targetManifestSizeBytes;
+
+  private final Encoder<ManifestFile> manifestEncoder = Encoders.javaSerialization(ManifestFile.class);
+  private final Encoder<Entry> entryEncoder = Encoders.javaSerialization(Entry.class);
+  private final Encoder<Bin> binEncoder = Encoders.bean(Bin.class);
+
+  private Predicate<ManifestFile> predicate = manifest -> true;
+  private String stagingLocation = null;
+
+  RewriteManifestsAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.specs = table.specs();
+    this.summary = Maps.newHashMap();
+    this.defaultParallelism = Integer.parseInt(
+        spark.conf().get("spark.default.parallelism", "200"));
+    this.snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+        table.properties(),
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
+    Preconditions.checkArgument(stagingLocation != null, "Staging location must be set");
+
+    List<ManifestFile> matchingManifests = findMatchingManifests();
+    if (matchingManifests.isEmpty()) {
+      return null;
+    }
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+
+    int parallelism = Math.min(matchingManifests.size(), defaultParallelism);
+    JavaRDD<ManifestFile> manifestRDD = sparkContext.parallelize(matchingManifests, parallelism);
+    Dataset<ManifestFile> manifestDS = spark.createDataset(manifestRDD.rdd(), manifestEncoder);
+    Dataset<Entry> manifestEntryDS = manifestDS.flatMap(toEntries(io, specs), entryEncoder);
+
+    try {
+      manifestEntryDS.cache();
```

Comment:
We never recommend caching and instead recommend using the shuffle system. Basically all you need to do is add a `repartition(parallelism)` and then run a `map` that does nothing. Then future executions will pick up the repartition's shuffle data.

That's much better than caching for clusters that use dynamic allocation:
* It doesn't depend on the application being scaled up to keep cache balanced
* It doesn't keep executors around for longer than needed
* It has a smaller memory requirement (which we can't control)

Reply:
I think both approaches make sense depending on circumstances. The data set will be relatively small and caching might be not a bad option without dynamic allocation as it won't require shuffling and writing/reading from disk. I would expose a config option so that people can pick whatever is more reasonable for their case. One point is clear: we should read metadata once.